### PR TITLE
Fixing panic issue in Validate when properties are nil

### LINKED ISSUE #1820: API model validation panics if `properties` is nil/empty
**Is this a request for help?**:
Yes
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE
---

**What version of acs-engine?**:
ALL
---

If caller sends a PUT request with a missing/omitted `properties` field ACS Engine validation panics:

E.g. of panic in: github.com\Azure\acs-engine\pkg\api\agentPoolOnlyApi\v20170831\validate.go (but happens for all API verisons)

Code snippet that panics:
    if e := validate.Struct(a); e != nil {
        return handleValidationErrors(e.(validator.ValidationErrors))
    }

Further panic happens because of the following line but the code above only handles ValidationErrors and not InvalidValidationErrors. This is because `properties` is nil so it's not a valid struct:

    if val.Kind() != reflect.Struct || val.Type() == timeType {
        return &InvalidValidationError{Type: reflect.TypeOf(s)}
    }
E.g. of sample PUT request:

        {
            "type": "Microsoft.ContainerService/managedClusters",
            "name": "[parameters('clusterName')]",
            "apiVersion": "2017-08-31",
            "location": "[resourceGroup().location]"
        }
 
Error message:
`{ "code": "InternalOperationError", "message": "err: 'interface conversion: error is *validator.InvalidValidationError, not validator.ValidationErrors' stacktrace: 'goroutine 1990793 [running]:\nruntime/debug.Stack(0xc4209aad38, 0x556b1503af00, 0xc4201824c0)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x7b\nacsrp/pkg/microsoft.com/containerservice/filter.ErrorHandling.func1(0xc420b93680)\n\t/tmp/go/src/acsrp/pkg/microsoft.com/containerservice/filter/errorhandling.go:28 +0x64\npanic(0x556b1503af00, 0xc4201824c0)\n\t/usr/local/go/src/runtime/panic.go:489 +0x2d3\nacsrp/vendor/github.com/Azure/acs-engine/pkg/api/agentPoolOnlyApi/v20170831.(*Properties).Validate(0x0, 0x15, 0x200)\n\t/tmp/go/src/acsrp/vendor/github.com/Azure/acs-engine/pkg/api/agentPoolOnlyApi/v20170831/validate.go:68 +0xc0\nacsrp/vendor/github.com/Azure/acs-engine/pkg/api.(*Apiloader).LoadContainerServiceForAgentPoolOnlyCluster(0xc4209ab148, 0xc4200f8c00, 0x15, 0x200, 0xc4207205e7, 0xa, 0x1, 0x9, 0x0, 0x0)\n\t/tmp/go/src/acsrp/vendor/github.com/Azure/acs-engine/pkg/api/apiloader.go:163 +0x220\nacsrp/pkg/microsoft.com/containerservice/server/operations/managedcluster.(*PutManagedClusterOperation).ValidateAndSaveGoalState(0xc4213120f0)\n\t/tmp/go/src/acsrp/pkg/microsoft.com/containerservice/server/operations/managedcluster/putmanagedcluster.go:95 +0xb6e\nacsrp/pkg/microsoft.com/containerservice/server/handlers.(*BaseHandler).ValidateAndPersistOperation(0xc42000e160, 0xc4204dcce0, 0x556b15a9a140, 0xc4213120f0)\n\t/tmp/go/src/acsrp/pkg/microsoft.com/containerservice/server/handlers/basehandler.go:31 +0x57\nacsrp/pkg/microsoft.com/containerservice/server/handlers.(*ManagedClusterManager).PutManagedClusterHandler(0xc42000e160, 0xc420820240, 0xc420b93680)\n\t/tmp/go/src/acsrp/pkg/microsoft.com/containerservice/server/handlers/managedclustermanager.go:69 +0x215\nacsrp/pkg/microsoft.com/containerservice/server/handlers.(*ManagedClusterManager).PutManagedClusterHandler-fm(0xc420820240, 0xc420b93680)\n\t/tmp/go/src/acsrp/cmd/containerservice/containerservice.go:502 +0x40\nacsrp/vendor/github.com/emicklei/go-restful.(*Container).dispatch.func6(0xc420820240, 0xc420b93680)\n\t/tmp/go/src/acsrp/vendor/github.com/emicklei/go-restful/container.go:267 +0x40\nacsrp/vendor/github.com/emicklei/go-restful.(*FilterChain).ProcessFilter(0xc420820330, 0xc420820240, 0xc420b93680)\n\t/tmp/go/src/acsrp/vendor/github.com/emicklei/go-restful/filter.go:21 +0x97\nacsrp/pkg/microsoft.com/containerservice/filter.ErrorHandling(0xc420820240, 0xc420b93680, 0xc420820330)\n\t/tmp/go/src/acsrp/pkg/microsoft.com/containerservice/filter/errorhandling.go:39 +0x67\nacsrp/vendor/github.com/emicklei/go-restful.(*FilterChain).ProcessFilter(0xc420820330, 0xc420820240, 0xc420b93680)\n\t/tmp/go/src/acsrp/vendor/github.com/emicklei/go-restful/filter.go:19 +0x6a\nacsrp/pkg/microsoft.com/containerservice/filter.Context(0xc420820240, 0xc420b93680, 0xc420820330)\n\t/tmp/go/src/acsrp/pkg/microsoft.com/containerservice/filter/context.go:43 +0x3af\nacsrp/vendor/github.com/emicklei/go-restful.(*FilterChain).ProcessFilter(0xc420820330, 0xc420820240, 0xc420b93680)\n\t/tmp/go/src/acsrp/vendor/github.com/emicklei/go-restful/filter.go:19 +0x6a\nacsrp/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc4200f0000, 0x556b15aa1f80, 0xc420562e00, 0xc420383c00)\n\t/tmp/go/src/acsrp/vendor/github.com/emicklei/go-restful/container.go:269 +0x9c8\nacsrp/vendor/github.com/emicklei/go-restful.(*Container).(acsrp/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x556b15aa1f80, 0xc420562e00, 0xc420383c00)\n\t/tmp/go/src/acsrp/vendor/github.com/emicklei/go-restful/container.go:120 +0x4a\nnet/http.HandlerFunc.ServeHTTP(0xc4209b7170, 0x556b15aa1f80, 0xc420562e00, 0xc420383c00)\n\t/usr/local/go/src/net/http/server.go:1942 +0x46\nnet/http.(*ServeMux).ServeHTTP(0x556b15b470e0, 0x556b15aa1f80, 0xc420562e00, 0xc420383c00)\n\t/usr/local/go/src/net/http/server.go:2238 +0x132\nnet/http.serverHandler.ServeHTTP(0xc42009b290, 0x556b15aa1f80, 0xc420562e00, 0xc420383c00)\n\t/usr/local/go/src/net/http/server.go:2568 +0x94\nnet/http.(*conn).serve(0xc4209c08c0, 0x556b15aa3640, 0xc420461740)\n\t/usr/local/go/src/net/http/server.go:1825 +0x614\ncreated by net/http.(*Server).Serve\n\t/usr/local/go/src/net/http/server.go:2668 +0x2d0\n'" }
`

### LINKED ISSUE #2586: Panic on acs-engine 0.14.6  generate when adding "enableAggregatedAPIs" to propeties
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): ISSUE

---

**What version of acs-engine?**:

Release linux AMD 64 - acs-engine-v0.14.6
https://github.com/Azure/acs-engine/releases/download/v0.14.6/acs-engine-v0.14.6-linux-amd64.tar.gz

---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**:
Using ubuntu bash on windows 10, but almost certain that it would be reproduceable in other shells
Downloaded acs engine like this:

```bash
curl -L https://github.com/Azure/acs-engine/releases/download/v0.14.6/acs-engine-v0.14.6-linux-amd64.tar.gz | tar zx
cp acs-engine-v0.14.6-linux-amd64/acs-engine .
chmod +x acs-engine
```

Used the following json file:

```json
{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.10",
      "kubernetesConfig": {
          "enableAggregatedAPIs": true
      }

    },
    "masterProfile": {
      "count": 3,
      "dnsPrefix": "wink8s001cluster",
      "vmSize": "Standard_DS2_v2"
    },
    "agentPoolProfiles": [
      {
        "name": "winagents",
        "count": 3,
        "vmSize": "Standard_D2s_v3",
        "availabilityProfile": "AvailabilitySet",
        "osType": "Windows"
      }
    ],
    "windowsProfile": {
      "adminUsername": "azureuser",
      "adminPassword": "4p4ssw0rd!"
    },
    "linuxProfile": {
      "adminUsername": "azureuser",
      "ssh": {
        "publicKeys": [
          {
            "keyData": ""
          }
        ]
      }
    },
    "servicePrincipalProfile": {
      "clientId": "",
      "secret": ""
    }
  }
}
```

You run
```
./acs-engine generate cluster.json
```

And you get the following error:

```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x6152f6]

goroutine 1 [running]:
github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver.(*Version).Prerelease(...)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver/version.go:146
github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver.constraintLessThan(0x0, 0xc420494980, 0x1)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver/constraints.go:256 +0x26
github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver.(*constraint).check(0xc420494980, 0x0, 0xc420734d80)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver/constraints.go:159 +0x38
github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver.Constraints.Check(0xc4206fc5a0, 0x1, 0x1, 0x0, 0x0)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/Masterminds/semver/constraints.go:49 +0x75
github.com/Azure/acs-engine/pkg/api/vlabs.(*OrchestratorProfile).Validate(0xc420050d40, 0x17bae00, 0xc42073af00, 0x0)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/pkg/api/vlabs/validate.go:94 +0x830
github.com/Azure/acs-engine/pkg/api/vlabs.(*Properties).Validate(0xc42073af00, 0x600, 0x8d8, 0xc420795960)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/pkg/api/vlabs/validate.go:310 +0xe3
github.com/Azure/acs-engine/pkg/api.(*Apiloader).LoadContainerService(0xc420795bc0, 0xc4200da900, 0x6d8, 0x8d8, 0xc4206bb1a8, 0x5, 0x1, 0x0, 0x19d9488, 0xc420795a20, ...)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/pkg/api/apiloader.go:170 +0x6f4
github.com/Azure/acs-engine/pkg/api.(*Apiloader).DeserializeContainerService(0xc420795bc0, 0xc4200da900, 0x6d8, 0x8d8, 0x1, 0x0, 0x0, 0x7fe233c30000, 0x0, 0x4979a9, ...)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/pkg/api/apiloader.go:44 +0x150
github.com/Azure/acs-engine/pkg/api.(*Apiloader).LoadContainerServiceFromFile(0xc420795bc0, 0x7fffe2700c25, 0xe, 0xc420770001, 0x0, 0x0, 0xc420778660, 0x19437ba, 0x5, 0xc4206eb5a8)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/pkg/api/apiloader.go:33 +0x211
github.com/Azure/acs-engine/cmd.(*generateCmd).validate(0xc4203af490, 0xc420774b40, 0xc420734600, 0x1, 0x1, 0x19429ce, 0x4)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/cmd/generate.go:98 +0x182
github.com/Azure/acs-engine/cmd.newGenerateCmd.func1(0xc420774b40, 0xc420734600, 0x1, 0x1, 0x0, 0x0)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/cmd/generate.go:48 +0x61
github.com/Azure/acs-engine/vendor/github.com/spf13/cobra.(*Command).execute(0xc420774b40, 0xc4207345d0, 0x1, 0x1, 0xc420774b40, 0xc4207345d0)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/spf13/cobra/command.go:647 +0x3f1
github.com/Azure/acs-engine/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420774480, 0xc4207758c0, 0xc420775680, 0xc420775200)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/spf13/cobra/command.go:726 +0x2fe
github.com/Azure/acs-engine/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420774480, 0xc420084008, 0x0)
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/vendor/github.com/spf13/cobra/command.go:685 +0x2b
main.main()
        /Users/jackfrancis/work/src/github.com/Azure/acs-engine/main.go:12 +0x74
```

**What you expected to happen**:

Give me a proper error message that it doesn't support  `enableAggregatedAPIs`, and not list @jackfrancis directories.

**How to reproduce it** (as minimally and precisely as possible):

Follow steps above.

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by tariq1890**
**What this PR does / why we need it**: Avoiding panics during validation

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2586 fixes #1820

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [x] unit tests
- [x] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/api/v20160330/validate.go [hunk 1]
```diff
@@ -65,6 +65,9 @@ func (l *LinuxProfile) Validate() error {
 
 // Validate implements APIObject
 func (a *Properties) Validate() error {
+	if a == nil {
```
Comment (jackfrancis) [inline pkg/api/v20160330/validate.go]: Can you point to some references that suggest that this is idiomatic golang? I think we prefer to do this nil-type validation before the method is invoked.

Comment (tariq1890) [reply pkg/api/v20160330/validate.go]: Oh yes, I did think of that initially. But I decided against it as I supposed this it'd be more consistent to put it within the Validate method. I'll make the change and push it.

**Comment by tariq1890** [comment]
@jackfrancis I've made the changes. As for the references, I am not sure if adding multiple comments is a good idea. I thought of pointing to this as a reference for other readers , https://golang.org/doc/faq#nil_error. Thoughts?

**Comment by CecileRobertMichon** [comment]
@shrutir25 @amanohar @weinong how does this PR align with #3197 ?

**Pull request merged**